Testing/Unit tests.md

```diff
@@ -1,29 +1,99 @@
 
-```c#
-[Fact]
-public async Task Process_WhenParameterIsNotValid_ThenThrowArgumentException()
-{ }
+Their simplicity makes running all unit tests in a project quick and easy, which is why we should be checking them periodically during development time and especially before publishing our branch and creating a PR. We can go even further with that idea. By integrating test runs in the CI/CD pipelines and PR merge requirements we cannot forget to run them before sharing the code with others!
+
+Detailed unit tests help us validate that our code snippets do what they were meant to do, but they won't validate that we've implemented the requested functionalities successfully. To do that, all our units must work together, which is why we have other types of tests like [integration](integration-tests) and [UI](ui-tests) tests.
+
+### Tips and tricks
+
+#### Testing the current time
+
+Let's say that we have to write unit tests for the following method:
+
+``` c#
+public string GetTimeBasedGreeting() 
+{
+    var now = DateTime.UtcNow;
+
+    if (now.Hour >= 6 && now.Hour < 12)
+        return "Good morning";
+    
+    if (now.Hour >= 12 && now.Hour < 18)
+        return "Good afternoon";
+
+    return "Good evening";
+}
 ```
 
-The most used test method is the Fact method which represents one single test with no parameters. But if you want to write tests with parameters and need to test the same assertion for multiple different inputs you can always use xUnit's Theory tests. There are multiple ways to pass the data but the most common is the InlineData. 
+It is easy to think of the test cases for this: we must check for each time of the day that the correct greeting is returned. But as soon as we start writing the tests, we come to an issue: only one greeting can be tested at a certain point in time. This means that we have no way of defining unit tests that will pass every time we run them, even though our code is working perfectly!
 
-```c#
-[Theory]
-[InlineData(-1, 1)]
-[InlineData(1, -1)]
-public async Task Process_WhenParameterIsNotValid_ThenThrowArgumentException(int a, int b)
+The issue here is our static reference to `DateTime.UtcNow`, which returns the current date and time. Unfortunately, despite our incredible technological advancements, we still haven't figured out how to control time. This means that as long as our code is dependent on something as uncontrollable as the unforgiving passage of time, we cannot make it run deterministically.
+
+What we _can_ do is add a layer of abstraction between our code and the way we fetch the current time. This can be done by adding an interface that we can easily mock when setting up our tests:
+
+``` c#
+public interface ITimeProvider
+{
+    public DateTime UtcNow { get; }
+}
```

Comment:
Just a thought, but would it be better to call it `IDateTimeProvider`? Just to use consistent naming since it is actually providing us `DateTime`

Reply:
I see your point, but I don't think we need to be that specific because it wouldn't work if we ever expanded the interface (e.g. by adding a property returning `DateTimeOffset` or any other time-related type). This example doesn't have that, but this is the simplest use case I could think of. Even though you can understand the term "time" as the time of the day, in this context I was thinking of the more abstract definition of time. Does that perspective make sense?

Reply:
Maybe the message would be clearer if I added some text about expanding the interface in the future?

Reply:
@icottiero agreed with your suggestion, I think we can leave it as is and just add a brief message about expanding the interface 😄